### DISCUSSION TITLE
How to save anomaly map images separately

### DISCUSSION THREAD

**Discussion by ss880426** [open-edge-platform/anomalib#1602] [Q&A]
How does EfficientAD generate anomaly map during inference? 
Thanks

**Comment by willyfh** [comment]
Hi @ss880426 , as you may already know, EfficientAD generates an anomaly map during inference using three neural network models: a teacher, a student, and an autoencoder. First, it computes the squared distance between the student's output and the teacher's output, as well as the squared distance between another part of the student's output and the autoencoder's output. Subsequently, the student-teacher map and the student-autoencoder map are generated by averaging the distance values over the channel dimension of the corresponding computed outputs. The final anomaly map is then produced by combining the two maps through quantile-based normalization.

For a better understanding, you can refer to these sections of the code:
https://github.com/openvinotoolkit/anomalib/blob/main/src/anomalib/models/efficient_ad/torch_model.py#L290

**Comment by ss880426** [comment]
Thank you for your reply
But what I want to ask is that after executing inference.py, I can’t output pictures like the ones in the paper.

![image](https://github.com/openvinotoolkit/anomalib/assets/59278951/219f1748-97a7-4448-a80b-032aaed81d88)

**Comment by willyfh** [comment]
Actually, the anomaly map will be output by the model, as you can see from this line: https://github.com/openvinotoolkit/anomalib/blob/main/src/anomalib/models/efficient_ad/torch_model.py#L357

You can try creating your own script to perform inference using the exported model. Alternatively, you may want to check out this notebook, which includes a demonstration of how to perform inference and visualize the anomaly map:
https://github.com/openvinotoolkit/anomalib/tree/main/notebooks/000_getting_started

If you specifically want such a color map, you can use JET colormap: https://docs.opencv.org/4.x/d3/d50/group__imgproc__colormap.html#gga9a805d8262bcbe273f16be9ea2055a65ab3f207661ddf74511b002b1acda5ec09

**Comment by ss880426** [comment]
I use train.py for training, but the default seems to be lighting instead of torch.
How do I adjust?
Thanks

**Comment by samet-akcay** [comment]
Hi @ss880426, let's start from the beginning. I assume you have something like this in your config file :)

```yaml
visualization:
  show_images: False # show images on the screen
  save_images: True # save images to the file system
  log_images: True # log images to the available loggers (if any)
  image_save_path: null # path to which images will be saved
  mode: full # options: ["full", "simple"]
```

where `mode` means what sort out visualisation you want. `full` stacks all the predictions together, similar to what you are getting as an output image. This is handled by this [module](https://github.com/openvinotoolkit/anomalib/blob/main/src/anomalib/post_processing/visualizer.py) module. 

If you want to stick to using `train.py` without writing your own script, you could quickly modify [_visualize_full](https://github.com/openvinotoolkit/anomalib/blob/main/src/anomalib/post_processing/visualizer.py#L139) function to customise your own figure to plot/show.

https://github.com/openvinotoolkit/anomalib/blob/470bb860d8478298fdf3f10a530184f8cdc9c938/src/anomalib/post_processing/visualizer.py#L139-L182

Let us know if this helps

**Comment by j84637587** [comment]
I use this code to export the anomaly map separately.
It's for debug usage, so it's ugly.

### v1.0.0+
Reference: https://github.com/openvinotoolkit/anomalib/blob/6be4c802396ce7aa393f89a60d141f739dc867e3/src/anomalib/models/image/efficient_ad/torch_model.py#L426
### v0.7.0-
Reference: https://github.com/openvinotoolkit/anomalib/blob/4abfa93dcfcb98771bc768b334c929ff9a02ce8b/src/anomalib/models/efficient_ad/torch_model.py#L350


After Changed:
```
            map_combined = 0.5 * map_st + 0.5 * map_stae  # <----- After this line

            fig, axes = plt.subplots(3, 8, figsize=(16, 4))
            for i in range(8):
                axes[0, i].imshow(map_st[i].cpu().detach().numpy().transpose(1, 2, 0))
                axes[0, i].axis("off")
                axes[0, i].set_title("Map Student")
                axes[1, i].imshow(map_stae[i].cpu().detach().numpy().transpose(1, 2, 0))
                axes[1, i].axis("off")
                axes[1, i].set_title("Map Student AE")
                axes[2, i].imshow(map_combined[i].cpu().detach().numpy().transpose(1, 2, 0))
                axes[2, i].axis("off")
                axes[2, i].set_title("Map Combined")
            plt.tight_layout()
            # save the plot or any method u prefer
```
